### PR TITLE
Move functorch python bindings to torch/csrc

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85430
* __->__ #85426

This moves functorch's python bindings to torch/csrc/functorch/init.cpp.
Coming next is the torchdim move. I didn't do torchdim yet because
moving functorch's python bindings unblocks some other things that I
want to do first.

Test Plan:
- tests

### CODE REVIEW

File: torch/csrc/functorch/init.cpp [hunk 1]
```diff
@@ -82,29 +85,6 @@ static std::pair<Tensor,int64_t> remove_existing_batch_dim(
   return std::make_pair(batched->value(), batched->bdim());
 }
 
-// Poor man's version of np.moveaxis. Moves the dimension at `dst` to `src`
-// while preserving the order of other existing dimensions.
-// We should probably add np.moveaxis (it is more general) to PyTorch. (#36048)
-// When we do, replace the following with it.
-static Tensor _movedim(const Tensor& self, int64_t src, int64_t dst) {
```
Comment (zou3519) [inline torch/csrc/functorch/init.cpp:89]: We have at::movedim already, so I replaced this with at::movedim.

File: torch/csrc/functorch/init.cpp [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+#include <ATen/functorch/LegacyVmapTransforms.h>
+#include <ATen/functorch/BatchedFallback.h>
+#include <ATen/functorch/BatchRulesHelper.h>
+#include <ATen/functorch/PlumbingHelper.h>
+#include <torch/csrc/functorch/CompileCache.h>
+#include <c10/core/AutogradState.h>
+#include <functorch/csrc/dim/dim.h>
+
+// This file contains functorch's Python bindings.
+
```
Comment (zou3519) [inline torch/csrc/functorch/init.cpp:23]: All of this code was moved from functorch/csrc/init.cpp. The only thing that changed were:
- namespace changes
- I deleted a (now obsolete) `_movedim` helper function.

File: torch/csrc/functorch/init.cpp [hunk 3]
```diff
@@ -282,8 +298,8 @@ static Tensor get_unwrapped(const Tensor& tensor) {
   if (wrapped) {
     return wrapped->value();
   }
-  auto* functional = dynamic_cast<FunctionalTensorWrapper*>(tensor.unsafeGetTensorImpl());
-  if (functional) {
+  if (at::functionalization::impl::isFunctionalTensor(tensor)) {
+    auto* functional = at::functionalization::impl::unsafeGetFunctionalWrapper(tensor);
```
Comment (zou3519) [inline torch/csrc/functorch/init.cpp:302]: dynamic_cast isn't portable (due to needing rtti) so I changed how we coerce a Tensor into a FunctionalTensorWrapper

File: torch/csrc/functorch/init.cpp [hunk 4]
```diff
@@ -4,25 +4,28 @@
...
+#include <torch/csrc/functorch/CompileCache.h>
 
 // This file contains functorch's Python bindings.
 
-namespace at {
+namespace torch {
 namespace functorch {
+namespace impl {
+
+using namespace at::functorch;
```
Comment (zou3519) [inline torch/csrc/functorch/init.cpp:28]: This file was moved, clang-format'ed, and namespaces were changed. The other changes are mentioned as inline comments on this PR.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**